### PR TITLE
Feature/complain

### PR DESCRIPTION
**Pull request by SamatB**

### CODE REVIEW

File: src/main/java/kg/giftlist/giftlistm2/controller/api/CharityController.java [hunk 1]
```diff
@@ -17,6 +17,7 @@
 @Tag(name = "Charity API", description = "User can get charity by id, get all charities, create, update or delete charity")
 @SecurityRequirement(name = "Authorization")
 @RequiredArgsConstructor
+@PreAuthorize("hasAuthority('USER')")
```
Comment (i-Coderr) [inline src/main/java/kg/giftlist/giftlistm2/controller/api/CharityController.java:20]: Remove from class and add it to respective user endpoints

Comment (i-Coderr) [reply src/main/java/kg/giftlist/giftlistm2/controller/api/CharityController.java:20]: I mean here not only user endpoints

File: src/main/java/kg/giftlist/giftlistm2/controller/api/CharityController.java [hunk 2]
```diff
@@ -64,4 +65,32 @@ public String deleteCharity(@PathVariable Long id) {
         return charityService.deleteCharity(id);
     }
 
+    @Operation(summary = "Admin gets a charity", description = "Admin can get a charity by id")
+    @GetMapping("charity/{id}")
```
Comment (i-Coderr) [inline src/main/java/kg/giftlist/giftlistm2/controller/api/CharityController.java:69]: api/charities/charity not looking good, maybe you will change it to api/charities/admin

File: src/main/java/kg/giftlist/giftlistm2/controller/api/CharityController.java [hunk 3]
```diff
@@ -64,4 +65,32 @@ public String deleteCharity(@PathVariable Long id) {
...
+    @PreAuthorize("hasAuthority('ADMIN')")
+    public CharityResponse getCharityByAdmin(@PathVariable Long id) {
+        return charityService.getCharityByAdmin(id);
+    }
+
+    @Operation(summary = "Admin gets all charities", description = "Admin can get all charities")
+    @GetMapping("charity")
```
Comment (i-Coderr) [inline src/main/java/kg/giftlist/giftlistm2/controller/api/CharityController.java:76]: Here also **admin**

File: src/main/java/kg/giftlist/giftlistm2/controller/api/ComplaintController.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+import java.util.List;
+
+@RestController
+@RequestMapping("api/complaints")
+@CrossOrigin
+@Tag(name = "Complaint API", description = "User can complain posts to admin, admin can delete a complain by id, block or unblock the post by id")
+@SecurityRequirement(name = "Authorization")
+@RequiredArgsConstructor
+@PreAuthorize("hasAuthority('ADMIN')")
```
Comment (i-Coderr) [inline src/main/java/kg/giftlist/giftlistm2/controller/api/ComplaintController.java:22]: Here also remove from class and add it to admin respective endpoints

File: src/main/java/kg/giftlist/giftlistm2/controller/api/WishListController.java [hunk 5]
```diff
@@ -52,22 +53,50 @@ public WishListResponse updateWishList(@PathVariable Long id,
...
     public String deleteWishList(@PathVariable Long id) {
         return wishListService.delete(id);
     }
 
-    @Operation(summary = "Wish list booking", description = "Booking a wish list")
-    @PostMapping("{id}")
-    public String booking(@PathVariable Long id) {
-        return wishListService.book(id);
+    @Operation(summary = "Admin gets a wish list", description = "Admin can get a wish list by id")
+    @GetMapping("wishlist/{id}")
```
Comment (i-Coderr) [inline src/main/java/kg/giftlist/giftlistm2/controller/api/WishListController.java:75]: Same api/whishlists/wishlist not looking good

File: src/main/java/kg/giftlist/giftlistm2/controller/api/WishListController.java [hunk 6]
```diff
@@ -52,22 +53,50 @@ public WishListResponse updateWishList(@PathVariable Long id,
...
+    public WishListResponse getWishlistByAdmin(@PathVariable Long id) {
+        return wishListService.getWishlistByAdmin(id);
     }
 
-    @Operation(summary = "Wish list un booking", description = "Un booking a wish list")
-    @DeleteMapping("wishlist/{id}")
-    public String unBook(@PathVariable Long id) {
-        return wishListService.unBook(id);
+    @Operation(summary = "Admin gets all wish lists", description = "Admin can get all wish lists")
+    @GetMapping("wishlist")
```
Comment (i-Coderr) [inline src/main/java/kg/giftlist/giftlistm2/controller/api/WishListController.java:82]: Same

File: src/main/java/kg/giftlist/giftlistm2/controller/payload/WishlistComplaintResponse.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+    private Long id;
+    private Long wishListId;
+    private String giftName;
+    private Long userId;
+    private String firstName;
+    private String lastName;
+    private String link;
+    private String holidayName;
+    private LocalDate holidayDate;
```
Comment (i-Coderr) [inline src/main/java/kg/giftlist/giftlistm2/controller/payload/WishlistComplaintResponse.java:21]: createdAt

File: src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    private Long userId;
+    private String firstName;
+    private String lastName;
+    private CharityStatus charityStatus;
+    private Condition condition;
+    private String category;
+    private String subcategory;
+    private String image;
+    private String description;
+    private LocalDate createdDate;
```
Comment (i-Coderr) [inline src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java:26]: createdAt

File: src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    private String complainingUserName;
```
Comment (i-Coderr) [inline src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java:27]: complainerFirstName

File: src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    private String complainingUserLastname;
```
Comment (i-Coderr) [inline src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java:28]: complainerLastName

File: src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    private String complaintCause;
```
Comment (i-Coderr) [inline src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java:29]: complaintReason

File: src/main/java/kg/giftlist/giftlistm2/db/entity/Booking.java [hunk 12]
```diff
@@ -41,4 +41,5 @@ public class Booking {
     @CreatedDate
     private LocalDate createdAt;
 
+
```
Comment (i-Coderr) [inline src/main/java/kg/giftlist/giftlistm2/db/entity/Booking.java:44]: Unnecessary line added

**Review by i-Coderr — COMMENTED** [review]
Commented at 10.32 pm 12/11/22

File: src/main/java/kg/giftlist/giftlistm2/controller/api/CharityController.java (hunk 1 above)
Comment (SamatB) [reply src/main/java/kg/giftlist/giftlistm2/controller/api/CharityController.java:20]: Done

File: src/main/java/kg/giftlist/giftlistm2/controller/api/CharityController.java (hunk 2 above)
Comment (SamatB) [reply src/main/java/kg/giftlist/giftlistm2/controller/api/CharityController.java:69]: Done

File: src/main/java/kg/giftlist/giftlistm2/controller/api/CharityController.java (hunk 3 above)
Comment (SamatB) [reply src/main/java/kg/giftlist/giftlistm2/controller/api/CharityController.java:76]: Done

File: src/main/java/kg/giftlist/giftlistm2/controller/api/ComplaintController.java (hunk 4 above)
Comment (SamatB) [reply src/main/java/kg/giftlist/giftlistm2/controller/api/ComplaintController.java:22]: Done

File: src/main/java/kg/giftlist/giftlistm2/controller/api/WishListController.java (hunk 5 above)
Comment (SamatB) [reply src/main/java/kg/giftlist/giftlistm2/controller/api/WishListController.java:75]: Done

File: src/main/java/kg/giftlist/giftlistm2/controller/api/WishListController.java (hunk 6 above)
Comment (SamatB) [reply src/main/java/kg/giftlist/giftlistm2/controller/api/WishListController.java:82]: Done

File: src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java (hunk 8 above)
Comment (SamatB) [reply src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java:26]: Done

File: src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java (hunk 9 above)
Comment (SamatB) [reply src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java:27]: Done

File: src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java (hunk 10 above)
Comment (SamatB) [reply src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java:28]: Done

File: src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java (hunk 11 above)
Comment (SamatB) [reply src/main/java/kg/giftlist/giftlistm2/controller/payload/CharityComplaintResponse.java:29]: Done

File: src/main/java/kg/giftlist/giftlistm2/controller/payload/WishlistComplaintResponse.java (hunk 7 above)
Comment (SamatB) [reply src/main/java/kg/giftlist/giftlistm2/controller/payload/WishlistComplaintResponse.java:21]: Done

File: src/main/java/kg/giftlist/giftlistm2/db/entity/Booking.java (hunk 12 above)
Comment (SamatB) [reply src/main/java/kg/giftlist/giftlistm2/db/entity/Booking.java:44]: Done

**Review by SamatB — COMMENTED** [review]
done

File: src/main/java/kg/giftlist/giftlistm2/mapper/FriendMapper.java [hunk 13]
```diff
@@ -28,17 +37,17 @@ public FriendProfileResponse friendResponse(User user) {
         friendProfileResponse.shoeSize(user.getShoeSize());
         friendProfileResponse.hobbies(user.getHobbies());
         friendProfileResponse.importantToKnow(user.getImportantToKnow());
-        List<WishList> wishLists = user.getWishLists();
+        List<WishList> wishLists = wishListRepository.getWishLists(user.getId());
         if (wishLists != null) {
-            friendProfileResponse.wishLists(new ArrayList<WishList>(wishLists));
```
Comment (SamatB) [inline src/main/java/kg/giftlist/giftlistm2/mapper/FriendMapper.java:33]: new ArrayList<WishList> and others below were deleted by Timurlan

Comment (i-Coderr) [reply src/main/java/kg/giftlist/giftlistm2/mapper/FriendMapper.java:33]: ok

**Review by i-Coderr — COMMENTED** [review]
Simply the best!

**Pull request merged**